routing/dht/providers/providers.go

```diff
@@ -0,0 +1,363 @@
+
+func mkProvKey(k key.Key) ds.Key {
+	return ds.NewKey(providersKeyPrefix + base64.RawURLEncoding.EncodeToString([]byte(k)))
+}
+
+func (pm *ProviderManager) Process() goprocess.Process {
+	return pm.proc
+}
+
+func (pm *ProviderManager) providersForKey(k key.Key) ([]peer.ID, error) {
+	pset, err := pm.getProvSet(k)
+	if err != nil {
+		return nil, err
+	}
+	return pset.providers, nil
+}
+
+func (pm *ProviderManager) getProvSet(k key.Key) (*providerSet, error) {
+	cached, ok := pm.providers.Get(k)
+	if ok {
+		return cached.(*providerSet), nil
+	}
+
+	pset, err := loadProvSet(pm.dstore, k)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(pset.providers) > 0 {
+		pm.providers.Add(k, pset)
+	}
+
+	return pset, nil
+}
+
+func loadProvSet(dstore ds.Datastore, k key.Key) (*providerSet, error) {
+	res, err := dstore.Query(dsq.Query{Prefix: mkProvKey(k).String()})
+	if err != nil {
+		return nil, err
+	}
+
+	out := newProviderSet()
+	for e := range res.Next() {
+		if e.Error != nil {
+			log.Error("got an error: ", err)
+			continue
+		}
+		parts := strings.Split(e.Key, "/")
+		if len(parts) != 4 {
+			log.Warning("incorrectly formatted key: ", e.Key)
+			continue
+		}
+
+		decstr, err := base64.RawURLEncoding.DecodeString(parts[len(parts)-1])
+		if err != nil {
+			log.Error("base64 decoding error: ", err)
+			continue
+		}
+
+		pid := peer.ID(decstr)
+
+		t, err := readTimeValue(e.Value)
+		if err != nil {
+			log.Warning("parsing providers record from disk: ", err)
+			continue
+		}
+
+		out.setVal(pid, t)
+	}
+
+	return out, nil
+}
+
+func readTimeValue(i interface{}) (time.Time, error) {
+	data, ok := i.([]byte)
+	if !ok {
+		return time.Time{}, fmt.Errorf("data was not a []byte")
+	}
+
+	nsec, _ := binary.Varint(data)
+
+	return time.Unix(0, nsec), nil
+}
+
+func (pm *ProviderManager) addProv(k key.Key, p peer.ID) error {
+	iprovs, ok := pm.providers.Get(k)
+	if !ok {
+		iprovs = newProviderSet()
+		pm.providers.Add(k, iprovs)
+	}
+	provs := iprovs.(*providerSet)
+	now := time.Now()
+	provs.setVal(p, now)
+
+	return writeProviderEntry(pm.dstore, k, p, now)
+}
+
+func writeProviderEntry(dstore ds.Datastore, k key.Key, p peer.ID, t time.Time) error {
+	dsk := mkProvKey(k).ChildString(base64.RawURLEncoding.EncodeToString([]byte(p)))
+
+	buf := make([]byte, 16)
+	n := binary.PutVarint(buf, t.UnixNano())
+
+	return dstore.Put(dsk, buf[:n])
+}
+
+func (pm *ProviderManager) deleteProvSet(k key.Key) error {
+	pm.providers.Remove(k)
+
+	res, err := pm.dstore.Query(dsq.Query{
+		KeysOnly: true,
+		Prefix:   mkProvKey(k).String(),
+	})
+
+	entries, err := res.Rest()
+	if err != nil {
+		return err
+	}
+
+	for _, e := range entries {
+		err := pm.dstore.Delete(ds.NewKey(e.Key))
+		if err != nil {
+			log.Error("deleting provider set: ", err)
+		}
+	}
+	return nil
+}
+
+func (pm *ProviderManager) getAllProvKeys() ([]key.Key, error) {
+	res, err := pm.dstore.Query(dsq.Query{
+		KeysOnly: true,
+		Prefix:   providersKeyPrefix,
+	})
+
```

Comment:
@whyrusleeping, using dstore.Query will work, but it is very slow.  The Query mechanism has a lot of overhead.  The data is being sent to you via an unbuffered channel in a separate go routine. In my own filestore code I was able to get a modest speedup by increasing the channel buffer size used the query, but it was still slow (speed up of 2 to 3x).  By querying the leveldb directly I was able to get a 10-12x speedup when doing a `filestore ls`.  (See ipfs-filestore/go-ipfs#10) I do not know how bad it will hurt performance here but it is something to keep it mind if you notice a slowdown after they code is deployed.